internal/storage/v2/memory/memory.go

```diff
@@ -0,0 +1,204 @@
+	}
+	return nil
+}
+
+// reshuffleScopeSpanAndSaveServiceAndOperation reshuffles all the scope spans of a resource span to group
+// spans of same trace ids together. The first step is to iterate the scope spans and then.
+// copy the scope to an empty scopeSpan. After this, append the spans with same
+// trace id to this empty scope span. Finally move this scope span to the scope span
+// slice containing other scope spans and having same trace id.
+func reshuffleScopeSpanAndSaveServiceAndOperation(m *Tenant, resourceSpan ptrace.ResourceSpans, sameTraceIDScopeSpans map[pcommon.TraceID]ptrace.ScopeSpansSlice) {
+	serviceName := getServiceNameFromResource(resourceSpan.Resource())
+	m.services[serviceName] = struct{}{}
+	for j := 0; j < resourceSpan.ScopeSpans().Len(); j++ {
+		sameTraceIDSpans := make(map[pcommon.TraceID]ptrace.SpanSlice)
+		scopeSpan := resourceSpan.ScopeSpans().At(j)
+		reshuffleSpansAndSaveOperation(m, serviceName, scopeSpan, sameTraceIDSpans)
+		for traceId, sameTraceIDSpanSlice := range sameTraceIDSpans {
+			sameTraceIDScopeSpan := ptrace.NewScopeSpans()
+			scopeSpan.Scope().CopyTo(sameTraceIDScopeSpan.Scope())
+			sameTraceIDSpanSlice.MoveAndAppendTo(sameTraceIDScopeSpan.Spans())
+			if sameTraceIDScopeSpanSlice, ok := sameTraceIDScopeSpans[traceId]; ok {
+				sameTraceIDScopeSpan.MoveTo(sameTraceIDScopeSpanSlice.AppendEmpty())
+			} else {
+				scopeSpanSlice := ptrace.NewScopeSpansSlice()
+				sameTraceIDScopeSpan.MoveTo(scopeSpanSlice.AppendEmpty())
+				sameTraceIDScopeSpans[traceId] = scopeSpanSlice
+			}
+		}
+	}
+}
+
+func reshuffleSpansAndSaveOperation(m *Tenant, serviceName string, scopeSpan ptrace.ScopeSpans, sameTraceIDSpans map[pcommon.TraceID]ptrace.SpanSlice) {
+	for k := 0; k < scopeSpan.Spans().Len(); k++ {
+		span := scopeSpan.Spans().At(k)
+		operation := getOperationFromSpanNameAndKind(span.Name(), span.Kind())
+		saveOperationToStore(m, serviceName, operation)
+		if _, ok := m.traces[span.TraceID()]; !ok {
+			m.traces[span.TraceID()] = ptrace.NewTraces()
+			// if we have a limit, let's cleanup the oldest traces
+			if m.config.MaxTraces > 0 {
+				// we only have to deal with this slice if we have a limit
+				m.index = (m.index + 1) % m.config.MaxTraces
+				// do we have an item already on this position? if so, we are overriding it,
+				// and we need to remove from the map
+				if !m.ids[m.index].IsEmpty() {
+					delete(m.traces, m.ids[m.index])
+				}
+				// update the ring with the trace id
+				m.ids[m.index] = span.TraceID()
+			}
+		}
+		// Collect all the spans with same trace id within the same scope span sameTraceIDSpanSlice
+		if sameTraceIDSpanSlice, ok := sameTraceIDSpans[span.TraceID()]; ok {
+			span.CopyTo(sameTraceIDSpanSlice.AppendEmpty())
+		} else {
+			spanSlice := ptrace.NewSpanSlice()
+			span.CopyTo(spanSlice.AppendEmpty())
+			sameTraceIDSpans[span.TraceID()] = spanSlice
+		}
+	}
+}
+
+func saveOperationToStore(m *Tenant, serviceName string, operation v2api.Operation) {
```

Comment:
if this function is meant to be called while holding a lock on the tenant then say so in the comment. Why isn't this a method of the Tenant struct?
```suggestion
func storeOperation(m *Tenant, serviceName string, operation v2api.Operation) {
```

Reply:
Should we make it a part of struct and not use lock? Will that be fine (with a comment)